crates/matrix-sdk-crypto/src/olm/group_sessions/sender_data_finder.rs

```diff
@@ -0,0 +1,891 @@
+
+        if let Some((msk, msk_verified)) = maybe_msk_info {
+            // Yes: H (cross-signing key matches that used to sign the device info!)
+            // and: J (device info is verified by matching cross-signing key)
+
+            // Find the actual key within the MasterPubkey struct
+            if let Some(msk) = msk.get_first_key() {
+                // We have MXID and MSK for the user sending the to-device message.
+                // Decide the MSK trust level based on whether we have verified this user.
+                // Set the MXID, MSK and trust level in the session. Remove the device
+                // info and retries since we don't need them.
+                Ok(SenderData::SenderKnown {
+                    user_id: sender_device.user_id().to_owned(),
+                    msk,
+                    msk_verified,
+                })
+                // TODO: [drop the lock]
+            } else {
+                // Surprisingly, there was no key in the MasterPubkey. We did not expect this:
+                // treat it as if the device was not signed by this master key.
+                //
+                tracing::error!(
+                    "MasterPubkey for user {} does not contain any keys!",
+                    msk.user_id()
+                );
+
+                Ok(SenderData::DeviceInfo {
+                    device_keys: sender_device.as_device_keys().clone(),
+                    retry_details: SenderDataRetryDetails::retry_soon(),
+                    legacy_session: true, // TODO: change to false when retries etc. are done
+                })
+                // TODO: [drop the lock]
+            }
+        } else {
+            // No: Device was not signed by the known identity of the sender.
+            // (Or the signature was invalid. We don't know which unfortunately, so we can't
+            // bail out completely if the signature was invalid, as we'd like
+            // to.)
+            //
+            // Since we've already checked there are >1 signatures, we guess
+            // it was signed by a different identity of this user, so we should retry
+            // later, in case either the device info or the user's identity changes.
+            Ok(SenderData::DeviceInfo {
+                device_keys: sender_device.as_device_keys().clone(),
+                retry_details: SenderDataRetryDetails::retry_soon(),
+                legacy_session: true, // TODO: change to false when retries etc. are done
+            })
+
+            // TODO: Increment retry_count and set
+            // next_retry_time_ms per backoff algorithm; let the background job
+            // pick it up [drop the lock]
+        }
+    }
+
+    /// If sender_device is correctly cross-signed by user_identity, return
+    /// user_identity's master key and whether it is verified.
+    /// Otherwise, return None.
+    async fn msk_if_device_is_signed_by_user<'i>(
```

Comment:
If you had access to a `Device` then you could this method:

https://github.com/matrix-org/matrix-rust-sdk/blob/8d54bd92d1b85dc830ebbfe79c7ecf7cced08f59/crates/matrix-sdk-crypto/src/identities/device.rs#L355-L358

Also, the `Device` already contains our own user identity and the identity of the device owner, this would simplify things a bunch. Why can't we instead use the `Store` instead of the `CryptoStore` trait in the finder to get access to the higher level `Device` types?

Otherwise, this is duplicating a lot of code we already have. I imagine other functions might be simpler if we had access to the `Store` as well.

If you're asking yourself why we even have those separate types, it's to avoid a reference cycle without introducing weak references. The `Device` contains a reference to the `CryptoStore` which might contain a reference to the `ReadOnlyDevice`.

Reply:
We don't have access to a `Device` because we have (potentially) read the `DeviceKeys` from the JSON of the to-device event. Is there a way for us to create one? I am not sure what I would provide for `verification_machine`, `own_identity and `device_owner_identity`.

Reply:
Well we have access to the `Store`, then we can build a `Device` from the `DeviceData`/`DeviceKeys`: https://github.com/matrix-org/matrix-rust-sdk/blob/7a85b7abdcfc05881ccbddf54a2094273be90110/crates/matrix-sdk-crypto/src/store/mod.rs#L1154-L1159

The only downside to this would be that we're operating on a `Device` that isn't in the store, but that's already the case for the `DeviceData` you create.

Reply:
Done in 4fef0a6fca and 8f1a4b1b2b